## 1. Load ```scikit-learn``` diabetes dataset

In [1]:
from sklearn.datasets import load_diabetes

sk_diabetes = load_diabetes()
print(sk_diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

## 2. Split ```scikit-learn``` Data Into Training and Test Sets

In [2]:
from sklearn.model_selection import train_test_split

X = sk_diabetes.data
y = sk_diabetes.target

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=308)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(331, 10) (111, 10) (331,) (111,)


## 3. Training and Test Set R<sup>2</sup> for Lasso Model With ```scikit-learn``` Dataset

In [3]:
import numpy as np
from sklearn.linear_model import Lasso

lasso = Lasso().fit(X_train, y_train)

# performance/score of a regressor is evaluated using R^2.
# Train R^2
print("Trainng R^2: %f" % lasso.score(X_train, y_train))
# Test R^2
print("Test R^2: %f" % lasso.score(X_test, y_test))

# Number of coefficients != 0 will tell us how many features are being used
features_used = lasso.coef_ != 0
print("Number of features being used: %d" % np.sum(features_used))

all_features = list(sk_diabetes.feature_names)

print("Names of features being used:")
for feature, used in zip(all_features, features_used):
    if used:
        print(feature)

Trainng R^2: 0.392352
Test R^2: 0.343602
Number of features being used: 3
Names of features being used:
bmi
bp
s5


### 3. Answers
* Training R<sup>2</sup> = 0.392352
* Test R<sup>2</sup> = 0.343602
* No. of features used = 3 (out of 10)
* Names of features being used:
    * bmi
    * bp
    * s5

## 4. Read and Load Original Diabetes Data From File

In [4]:
# Read diabetes data from file.
# First line is headers so we dont include that with the data.

import numpy as np
diabetes = np.genfromtxt("diabetes.data", delimiter='\t')[1:]

y = diabetes[:, -1] # Target, last column.
X = diabetes[:, :-1] # Data, all columns except last.

print(diabetes.shape, X.shape, y.shape)

(442, 11) (442, 10) (442,)


## 5. Split Original Diabetes Data Into Training and Test Sets

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=308)

## 6. Training and Test Set R<sup>2</sup> for Lasso Model with Original Diabetes Dataset

In [6]:
from sklearn.linear_model import Lasso

lasso = Lasso().fit(X_train, y_train)

# performance/score of a regressor is evaluated using R^2.
# Train R^2
print("Trainng R^2: %f" % lasso.score(X_train, y_train))
# Test R^2
print("Test R^2: %f" % lasso.score(X_test, y_test))

# Number of coefficients != 0 will tell us how many features are being used
features_used = lasso.coef_ != 0
print("Number of features being used: %d" % np.sum(features_used))

print("Names of features being used:")
for feature, used in zip(all_features, features_used):
    if used:
        print(feature)

Trainng R^2: 0.532342
Test R^2: 0.423408
Number of features being used: 9
Names of features being used:
age
sex
bmi
bp
s2
s3
s4
s5
s6


### 6. Answers
* Training R<sup>2</sup> = 0.532342
* Test R<sup>2</sup> = 0.423408
* No. of features used = 9 (out of 10)
* Names of features being used:
    * age
    * sex
    * bmi
    * bp
    * s2
    * s3
    * s4
    * s5
    * s6
    
> The performance of lasso on this dataset is better than the one from the ```scikit-learn``` version, which can be see from the higher R<sup>2</sup> values for the training and test sets. We can also see that for the current dataset almost all of the features are being used, whereas very few were used in the previous one (9 out of 10 are being used for the current dataset, while only 3 out of 10 were used in the previous one).

## 7. Preprocess Training and Test Sets Using ```StandardScaler```

In [7]:
from sklearn.preprocessing import StandardScaler 

# Instantiate our scaler
scaler = StandardScaler()
scaler.fit(X_train)

# Apply same scaler to training and test set to preprocess them in the same way
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## 8. Training and Test Set R<sup>2</sup> for Lasso Model With Preprocessed Original Diabetes Dataset

In [8]:
from sklearn.linear_model import Lasso

lasso = Lasso().fit(X_train_scaled, y_train)

# performance/score of a regressor is evaluated using R^2.
# Train R^2
print("Trainng R^2: %f" % lasso.score(X_train_scaled, y_train))
# Test R^2
print("Test R^2: %f" % lasso.score(X_test_scaled, y_test))

# Number of coefficients != 0 will tell us how many features are being used
features_used = lasso.coef_ != 0
print("Number of features being used: %d" % np.sum(features_used))

print("Names of features being used:")
for feature, used in zip(all_features, features_used):
    if used:
        print(feature)

Trainng R^2: 0.534242
Test R^2: 0.419899
Number of features being used: 7
Names of features being used:
sex
bmi
bp
s1
s3
s4
s5


### 8. Answers
* Training R<sup>2</sup> = 0.534242
* Test R<sup>2</sup> = 0.419899
* No. of features used = 7 (out of 10)
* Names of features being used:
    * sex
    * bmi
    * bp
    * s1
    * s3
    * s4
    * s5
    
The performance of lasso on this preprocessed dataset is closer to that of [6] as opposed to [3]. This goes against the expectation that it would be closer to the results for the normalised data from [3]. This could be because this dataset and the dataset used in [3] were normalised in different ways. 

This dataset uses ```StandardScaler```, which makes the mean of each feature 0 and its standard deviation 1. The dataset from [3] however normalises it by doing the following (according to the DESCR):

> "Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1)."



## 9. Plot Test R<sup>2</sup> against number of features used for varying values of the α parameter

In [9]:
test_scores = []
features_used = []
alpha_values = [2 ** k for k in range(-5, 5)]

for a in alpha_values:
    
    lasso = Lasso(alpha=a).fit(X_train_scaled, y_train)
    test_score = lasso.score(X_test_scaled, y_test)
    features = np.sum(lasso.coef_ != 0)
    
    print("Alpha: %f . Test R^2: %f . Features used: %d " % (a, test_score, features))

    test_scores.append(test_score)
    features_used.append(features)

import matplotlib.pyplot as plt

fig = plt.figure(figsize=(17, 7))

for i in range(len(test_scores)):
    plt.scatter(test_scores[i], features_used[i], label=alpha_values[i])

plt.legend(loc='center left', bbox_to_anchor=(1, 0.5), title='alpha')

Alpha: 0.031250 . Test R^2: 0.419980 . Features used: 10 
Alpha: 0.062500 . Test R^2: 0.420708 . Features used: 10 
Alpha: 0.125000 . Test R^2: 0.421665 . Features used: 10 
Alpha: 0.250000 . Test R^2: 0.420145 . Features used: 10 
Alpha: 0.500000 . Test R^2: 0.419872 . Features used: 9 
Alpha: 1.000000 . Test R^2: 0.419899 . Features used: 7 
Alpha: 2.000000 . Test R^2: 0.416763 . Features used: 6 
Alpha: 4.000000 . Test R^2: 0.402941 . Features used: 5 
Alpha: 8.000000 . Test R^2: 0.396487 . Features used: 4 
Alpha: 16.000000 . Test R^2: 0.374059 . Features used: 3 


From the plot, we can see that the larger the α value, the poorer the performance (as we get smaller values for Test R<sup>2</sup>).

I had initially used α = 2<sup>k</sup> for k in [-5, -4, ..., 4, 5], however, k=5 gave an α of 32, with Test R<sup>2</sup> = 0.238489 and Features used = 2. This score was significantly lower than the others, so I did not use this value and stopped at k=4.

If we don't want to use all the features, a good α would be 2. I prefer this one because it's performance is close to that of others with a lower α, while it only uses 6 features(so it has a good balance).

## 10. Choose the regularisation parameter for the Lasso using cross-validation on the training set

By using ```GridSearchCV``` on Lasso with a range of values for the α parameter, we can find the best choice for α.

In [10]:
from sklearn.model_selection import GridSearchCV

param_grid = { 'alpha': [2 ** k for k in range(-5, 6)] }
grid_search = GridSearchCV(Lasso(), param_grid, cv=5, iid=True)

grid_search.fit(X_train_scaled, y_train)

print("Best parameters: ", grid_search.best_params_)
print("Best cross-validation accuracy: ", grid_search.best_score_)

# Gives the model with the best parameters, trained on the whole training set 
best_lasso = grid_search.best_estimator_

print("Trainng R^2: %f" % best_lasso.score(X_train_scaled, y_train))
print("Test R^2: %f" % best_lasso.score(X_test_scaled, y_test))

features_used = best_lasso.coef_ != 0
print("Number of features being used: %d" % np.sum(features_used))

print("Names of features being used:")
for feature, used in zip(all_features, features_used):
    if used:
        print(feature)

Best parameters:  {'alpha': 2}
Best cross-validation accuracy:  0.5098536654307827
Trainng R^2: 0.529743
Test R^2: 0.416763
Number of features being used: 6
Names of features being used:
sex
bmi
bp
s1
s3
s5


### 10. Answers
* Training R<sup>2</sup> = 0.529743
* Test R<sup>2</sup> = 0.416763
* No. of features used = 6 (out of 10)
* Names of features being used:
    * sex
    * bmi
    * bp
    * s1
    * s3
    * s5

## 11. Implement an Inductive Conformal Predictor

> a) Split the training set that you obtained in item **5** into two parts: the ***calibration set of size 99*** and the ***rest of the training set*** (the ***training set proper***).

In [11]:
# X_train and y_train are the training sets from item 5
X_train_pr, X_calibration, y_train_pr, y_calibration = train_test_split(X_train, y_train, test_size=99, random_state=308)

print(X_calibration.shape, X_train_pr.shape, y_calibration.shape, y_train_pr.shape)


(99, 10) (232, 10) (99,) (232,)


> b) Preprocess the ***training set proper, calibration set,*** and ***test set*** in the same way using ```StandardScaler```.

In [12]:
# Instantiate our scaler
s = StandardScaler()

# Fit the scaler to the training set proper
s.fit(X_train_pr)

# Apply same scaler to training set proper, calibration set, and test set to preprocess them in the same way
X_train_pr_scaled = s.transform(X_train_pr)
X_calibration_scaled = s.transform(X_calibration)
X_test_scaled = s.transform(X_test)

> c) Using the nonconformity measure α = |y − yˆ|, where y is the true label and yˆ is its prediction given the training set proper, for each test sample compute the prediction interval for it. Do this for significance levels 5% and 20%. For each of these significance levels compute:
* the length of the prediction intervals for the test samples
* and the test error rate of your inductive conformal predictor.

In [13]:
# Fit lasso to training set proper
lasso = Lasso(alpha=2).fit(X_train_pr_scaled, y_train_pr)

# Get the lasso predictions for our calibration set
predicted_values_calibration = lasso.predict(X_calibration_scaled)

# Compute α = |y − yˆ|, where y is the true label and yˆ is its prediction.
non_conformity_scores = [ abs(true - pred) 
                         for true, pred 
                         in zip(y_calibration, predicted_values_calibration) ]


sorted_non_conformities = non_conformity_scores.copy()
sorted_non_conformities.sort()

from math import ceil

# Calculate k and c for 5% and 20% significance level

k5 = ceil( (1 - 0.05) * (len(X_calibration_scaled) + 1) )
c5 = sorted_non_conformities[k5-1]

k20 = ceil( (1 - 0.2) * (len(X_calibration_scaled) + 1) )
c20 = sorted_non_conformities[k20-1]

predicted_values_test = lasso.predict(X_test_scaled)

prediction_intervals_5 = []
prediction_intervals_20 = []
correct_5 = []
correct_20 = []

for sample, true_label, predicted_label in zip(X_test_scaled, y_test, predicted_values_test):
    
    # Calculate prediction interval for test sample
    pred_interval_5 = [ predicted_label - c5, predicted_label + c5 ]
    pred_interval_20 = [ predicted_label - c20, predicted_label + c20 ]
    
    prediction_intervals_5.append(pred_interval_5)
    prediction_intervals_20.append(pred_interval_20)
    
    # Inductive conformal predictor is only correct if true label is coved by prediction interval.
    prediction_correct_5 = pred_interval_5[0] <= true_label <= pred_interval_5[1]
    prediction_correct_20 = pred_interval_20[0] <= true_label <= pred_interval_20[1]
    
    correct_5.append(prediction_correct_5)
    correct_20.append(prediction_correct_20)
    
    
# Using our definition of a predition set, all prediction sets are of the same length. 
# This means we can check the length of any of the prediction intervals.
length_interval_5 = np.diff(prediction_intervals_5[0])
length_interval_20 = np.diff(prediction_intervals_20[0])

# Test error rate = number of errors made / total number of test samples
# Since the value we have is the number of correct prediction itervals, we subtract it from 1 to get the error rate.
test_error_rate_5 = 1 - np.mean(correct_5)
test_error_rate_20 = 1 - np.mean(correct_20)

print("5% Confidence level:")
print("\tInterval Length: %f" % length_interval_5)
print("\tTest Error Rate: %f" % test_error_rate_5)
print("20% Confidence level:")
print("\tInterval Length: %f" % length_interval_20)
print("\tTest Error Rate: %f" % test_error_rate_20)

5% Confidence level:
	Interval Length: 180.514327
	Test Error Rate: 0.117117
20% Confidence level:
	Interval Length: 117.188962
	Test Error Rate: 0.378378


# 12. Results

> ### a) The training and test R<sup>2</sup> for the Lasso model with default parameters on the ```scikit-learn``` version of diabetes and the number of features used.

Training R<sup>2</sup> = 0.392352

Test R<sup>2</sup> = 0.343602

No. of features used = 3 (out of 10)

Names of features being used:
* bmi
* bp
* s5

> ### b) The training and test R<sup>2</sup> for the Lasso model with default parameters on the original version of diabetes and the number of features used.


Training R<sup>2</sup> = 0.532342

Test R<sup>2</sup> = 0.423408

No. of features used = 9 (out of 10)

Names of features being used:
* age
* sex
* bmi
* bp
* s2
* s3
* s4
* s5
* s6
    
The performance of lasso on this dataset is better than the one from the ```scikit-learn``` version, which can be see from the higher R<sup>2</sup> values for the training and test sets. We can also see that for the current dataset almost all of the features are being used, whereas very few were used in the previous one (9 out of 10 are being used for the current dataset, while only 3 out of 10 were used in the previous one).

> ### c) The training and test R<sup>2</sup> for the Lasso model with default parameters on your version of diabetes (i.e., on the original version of diabetes with the features normalized by you); the number of features used.


Training R<sup>2</sup> = 0.534242

Test R<sup>2</sup> = 0.419899

No. of features used = 7 (out of 10)

Names of features being used:
* sex
* bmi
* bp
* s1
* s3
* s4
* s5

The performance of lasso on this preprocessed dataset is closer to that of [6] as opposed to [3]. This goes against the expectation that it would be closer to the results for the normalised data from [3]. This could be because this dataset and the dataset used in [3] were normalised in different ways. 

This dataset uses ```StandardScaler```, which makes the mean of each feature 0 and its standard deviation 1. The dataset from [3] however normalises it by doing the following (according to the DESCR):

> "Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1)."

> ### d) The training and test R<sup>2</sup> for the Lasso model with the best parameters chosen by cross-validation on your version of diabetes; the number of features used.


Training R<sup>2</sup> = 0.529743

Test R<sup>2</sup> = 0.416763

No. of features used = 6 (out of 10)

Names of features being used:
* sex
* bmi
* bp
* s1
* s3
* s5

> ### e) The lengths of prediction intervals and their test error rates at significance levels 5% and 20% (if you are implementing an inductive conformal predictor).


5% Confidence level:
* Interval Length: 180.514327
* Test Error Rate: 0.117117 ( = 11.71% )

20% Confidence level:
* Interval Length: 117.188962
* Test Error Rate: 0.378378 ( = 37.84% )

While 5% has a lower test error rate than 20%, it also has a greater interval length.